### PR TITLE
[spirv] Extend spv.array with Layoutinfo

### PR DESCRIPTION
**Pull request by denis0x0D**
This patch is addressing issue 
https://github.com/tensorflow/mlir/issues/76

1. The ebnf is updated to support layoutinfo:
For example, spv.array with ArrayStride information:
`!spv.array< 4 x !spv.array<4 x f32 [4]> [128]>`
 The check for layoutinfo in the parser is relaxed, the checks in the (de)serializer generate erros in case ArrayStride is not present for OpTypeArray regarding to the spec.
2. Tests related to serialization machinery updated.
3. Changed specialization of template function from `typename StructType::LayoutInfo_` to `typename uint64_t` to be able to reuse it for ArrayType::LayoutInfo.

@antiagainst @MaheshRavishankar can you please take a look?
Thanks!

### CODE REVIEW

File: lib/Dialect/SPIRV/SPIRVDialect.cpp [hunk 1]
```diff
@@ -150,7 +162,8 @@ static Type parseAndVerifyType(SPIRVDialect const &dialect, StringRef spec,
 //                | vector-type
 //                | spirv-type
 //
-// array-type ::= `!spv.array<` integer-literal `x` element-type `>`
+// array-type ::= `!spv.array<` integer-literal `x` element-type
+// (`[` integer-literal `]`)? `>`
```
Comment (MaheshRavishankar) [inline lib/Dialect/SPIRV/SPIRVDialect.cpp]: Nit : Please indent the expression to align with `!spv.array`

File: lib/Dialect/SPIRV/SPIRVDialect.cpp [hunk 2]
```diff
@@ -53,6 +53,18 @@ SPIRVDialect::SPIRVDialect(MLIRContext *context)
 // Type Parsing
 //===----------------------------------------------------------------------===//
 
+// Forward declarations.
+template <typename ValTy>
+Optional<ValTy> parseAndVerify(SPIRVDialect const &dialect, Location loc,
+                               StringRef spec);
+template <>
+Optional<Type> parseAndVerify<Type>(SPIRVDialect const &dialect, Location loc,
```
Comment (MaheshRavishankar) [inline lib/Dialect/SPIRV/SPIRVDialect.cpp]: I don't think we need to forward declare the template specialization.

File: lib/Dialect/SPIRV/SPIRVDialect.cpp [hunk 3]
```diff
@@ -267,18 +307,16 @@ Optional<Type> parseAndVerify<Type>(SPIRVDialect const &dialect, Location loc,
...
-parseAndVerify(SPIRVDialect const &dialect, Location loc, StringRef spec) {
+Optional<uint64_t> parseAndVerify(SPIRVDialect const &dialect, Location loc,
+                                  StringRef spec) {
   uint64_t offsetVal = std::numeric_limits<uint64_t>::max();
   if (!spec.consume_front("[")) {
     emitError(loc, "expected '[' while parsing layout specification in '")
         << spec << "'";
     return llvm::None;
   }
   if (spec.consumeInteger(10, offsetVal)) {
```
Comment (MaheshRavishankar) [inline lib/Dialect/SPIRV/SPIRVDialect.cpp]: This is not related to your change, but can you add spec.trim() before spec.consumeInteger?

File: lib/Dialect/SPIRV/Serialization/Deserializer.cpp [hunk 4]
```diff
@@ -207,6 +207,9 @@ class Deserializer {
   // Result <id> to decorations mapping.
   DenseMap<uint32_t, NamedAttributeList> decorations;
 
+  // Result <id> to type decorations.
+  DenseMap<uint32_t, uint64_t> typeDecorations;
```
Comment (MaheshRavishankar) [inline lib/Dialect/SPIRV/Serialization/Deserializer.cpp]: I think it might be better to make this DenseMap<uint32_t, uint32_t> cause it will be interpreted to different types later.

File: lib/Dialect/SPIRV/Serialization/Deserializer.cpp [hunk 5]
```diff
@@ -321,6 +324,9 @@ LogicalResult Deserializer::processDecoration(ArrayRef<uint32_t> words) {
         opBuilder.getIdentifier(attrName),
         opBuilder.getI32IntegerAttr(static_cast<int32_t>(words[2])));
     break;
+  case spirv::Decoration::ArrayStride:
+    typeDecorations[words[0]] = static_cast<uint64_t>(words[2]);
```
Comment (MaheshRavishankar) [inline lib/Dialect/SPIRV/Serialization/Deserializer.cpp]: Could you add a check to see that words.size() == 3 (like it is done for the other decoration types.

File: lib/Dialect/SPIRV/Serialization/Deserializer.cpp [hunk 6]
```diff
@@ -581,7 +587,15 @@ LogicalResult Deserializer::processArrayType(ArrayRef<uint32_t> operands) {
            << defOp->getName();
   }
 
-  typeMap[operands[0]] = spirv::ArrayType::get(elementTy, count);
+  spirv::ArrayType::LayoutInfo layoutInfo = 0;
```
Comment (MaheshRavishankar) [inline lib/Dialect/SPIRV/Serialization/Deserializer.cpp]: You could just do `spirv::ArrayType::LayoutInfo layoutInfo = typeDecorations.lookup(operands[0])`

File: lib/Dialect/SPIRV/Serialization/Serializer.cpp [hunk 7]
```diff
@@ -436,7 +461,15 @@ LogicalResult Serializer::processType(Location loc, Type type,
                                      operands))) ||
       succeeded(prepareBasicType(loc, type, typeEnum, operands))) {
     typeIDMap[type] = typeID;
-    return encodeInstructionInto(typesGlobalValues, typeEnum, operands);
+
+    if (failed(encodeInstructionInto(typesGlobalValues, typeEnum, operands))) {
+      return failure();
+    }
+
+    if (auto dType = type.dyn_cast<spirv::ArrayType>()) {
```
Comment (MaheshRavishankar) [inline lib/Dialect/SPIRV/Serialization/Serializer.cpp]: Why is this here? Shouldn't this be within `prepareBasicType` where it is checking for handling ArrayType?

File: lib/Dialect/SPIRV/SPIRVDialect.cpp (hunk 3 above)
Comment (denis0x0D) [reply lib/Dialect/SPIRV/SPIRVDialect.cpp]: I'll update it.

File: lib/Dialect/SPIRV/SPIRVDialect.cpp (hunk 2 above)
Comment (denis0x0D) [reply lib/Dialect/SPIRV/SPIRVDialect.cpp]: Thanks, I just didn't want to move the functions body, since I've used it in array parser, early then definitions. So, it it's ok to move them.

File: lib/Dialect/SPIRV/SPIRVDialect.cpp (hunk 1 above)
Comment (denis0x0D) [reply lib/Dialect/SPIRV/SPIRVDialect.cpp]: Thanks.

File: lib/Dialect/SPIRV/SPIRVDialect.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply lib/Dialect/SPIRV/SPIRVDialect.cpp]: I didn't mean move them. You can forward declare the template function. I dont think you need to forward declare the specialization. So keep 
```
template <typename ValTy>
Optional<ValTy> parseAndVerify(SPIRVDialect const &dialect, Location loc,
                               StringRef spec);
```

but remove the 

```
template <>
Optional<Type> parseAndVerify<Type>(SPIRVDialect const &dialect, Location loc, StringRef spec)
```
and 

```
template <>
Optional<uint64_t> parseAndVerify<uint64_t>(SPIRVDialect const &dialect, Location loc, StringRef spec);
```

I think that should work. If I am mistaken, I am sorry about that then.

Comment (denis0x0D) [reply lib/Dialect/SPIRV/SPIRVDialect.cpp]: Oh, sorry didn't get it right at first. Seems like in this case we will get an error like specilization of "function", after instantiation.

File: lib/Dialect/SPIRV/Serialization/Serializer.cpp (hunk 7 above)
Comment (denis0x0D) [reply lib/Dialect/SPIRV/Serialization/Serializer.cpp]: As far as I understood, please fix me if I'm wrong, to encode instruction with OpDecorate we need to call `encodeInstructionInto() `with last argument as array of args, when the first element of arg is unique id, so to process it in the prepareBasic function, we have to change the API prepareBasicType, so, if it's ok, I'll change it.
Thanks.

Comment (denis0x0D) [reply lib/Dialect/SPIRV/Serialization/Serializer.cpp]: Oh, I see the function prepareBasicType is private, so, if it's ok to I'll add one more args as resultId.

File: lib/Dialect/SPIRV/Serialization/Deserializer.cpp (hunk 5 above)
Comment (denis0x0D) [reply lib/Dialect/SPIRV/Serialization/Deserializer.cpp]: Thanks!

File: lib/Dialect/SPIRV/Serialization/Deserializer.cpp (hunk 6 above)
Comment (denis0x0D) [reply lib/Dialect/SPIRV/Serialization/Deserializer.cpp]: Thanks for mention this!

File: lib/Dialect/SPIRV/Serialization/Deserializer.cpp (hunk 4 above)
Comment (denis0x0D) [reply lib/Dialect/SPIRV/Serialization/Deserializer.cpp]: Thanks, I'll update it.

**Comment by denis0x0D** [comment]
@MaheshRavishankar  thanks a lot for review! I've updated the patch, can you please take a look?

File: lib/Dialect/SPIRV/SPIRVDialect.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply lib/Dialect/SPIRV/SPIRVDialect.cpp]: Oh Ok. Then sorry about that. This looks fine then.

File: lib/Dialect/SPIRV/Serialization/Serializer.cpp (hunk 7 above)
Comment (MaheshRavishankar) [reply lib/Dialect/SPIRV/Serialization/Serializer.cpp]: This looks good. Thanks!

**Comment by denis0x0D** [comment]
@MaheshRavishankar thanks a lot for review!

File: lib/Dialect/SPIRV/SPIRVDialect.cpp [hunk 8]
```diff
@@ -171,11 +184,38 @@ static Type parseArrayType(SPIRVDialect const &dialect, StringRef spec,
...
+
+  // Handle case when element type is not a trivial type
+  auto lastRDelimiter = spec.rfind('>');
+  if (lastRDelimiter != StringRef::npos) {
+    lastLSquare = spec.find('[', lastRDelimiter);
+  } else {
+    lastLSquare = spec.rfind('[');
+  }
+
+  // The check is relaxed in the parser
```
Comment (antiagainst) [inline lib/Dialect/SPIRV/SPIRVDialect.cpp]: What does this comment mean? Can we change it to something more suitable?

File: lib/Dialect/SPIRV/Serialization/Deserializer.cpp [hunk 9]
```diff
@@ -581,7 +591,12 @@ LogicalResult Deserializer::processArrayType(ArrayRef<uint32_t> operands) {
            << defOp->getName();
   }
 
-  typeMap[operands[0]] = spirv::ArrayType::get(elementTy, count);
+  spirv::ArrayType::LayoutInfo layoutInfo = typeDecorations.lookup(operands[0]);
+  if (!layoutInfo) {
```
Comment (antiagainst) [inline lib/Dialect/SPIRV/Serialization/Deserializer.cpp]: This is not true. Only types used by interface variables need to be explicitly laid out. Types used by function local variables are not. So we should drop this check.

File: lib/Dialect/SPIRV/Serialization/Serializer.cpp [hunk 10]
```diff
@@ -355,6 +359,27 @@ LogicalResult Serializer::processDecoration(Location loc, uint32_t resultID,
...
+    Location loc, spirv::ArrayType type, uint32_t resultID) {
+  if (type.hasLayout()) {
+    // OpDecorate %arrayTypeSSA ArrayStride strideLiteral
+    SmallVector<uint32_t, 3> args;
+    args.push_back(resultID);
+    args.push_back(static_cast<uint32_t>(spirv::Decoration::ArrayStride));
+    args.push_back(type.getArrayStride());
+    return encodeInstructionInto(decorations, spirv::Opcode::OpDecorate, args);
+  }
+  // Regarding to the spec each OpTypeArray must have ArrayStride
```
Comment (antiagainst) [inline lib/Dialect/SPIRV/Serialization/Serializer.cpp]: This is not true.

File: include/mlir/Dialect/SPIRV/SPIRVTypes.h [hunk 11]
```diff
@@ -73,14 +73,22 @@ class ArrayType : public Type::TypeBase<ArrayType, CompositeType,
                                         detail::ArrayTypeStorage> {
 public:
   using Base::Base;
+  using LayoutInfo = uint64_t;
```
Comment (antiagainst) [inline include/mlir/Dialect/SPIRV/SPIRVTypes.h]: Could you put a comment here to mention that if this is 0 it means no layout info?

File: lib/Dialect/SPIRV/Serialization/Deserializer.cpp (hunk 9 above)
Comment (denis0x0D) [reply lib/Dialect/SPIRV/Serialization/Deserializer.cpp]: Oh, sorry, my fault,
I was relying on this statement in the spec:

> **Each array type must have an ArrayStride decoration**, unless it is an array that contains a structure decorated with Block or BufferBlock, in which case it must not have an ArrayStride decoration.

and completely missed the header 

> Composite objects in the **StorageBuffer**, **Uniform**, and **PushConstant** Storage Classes must be explicitly laid out. The following apply to all the aggregate and matrix types describing such an object, recursively through their nested types:

Thanks a lot for fixing me!

File: lib/Dialect/SPIRV/SPIRVDialect.cpp (hunk 8 above)
Comment (denis0x0D) [reply lib/Dialect/SPIRV/SPIRVDialect.cpp]: Thanks, I've delete it.

File: lib/Dialect/SPIRV/Serialization/Serializer.cpp (hunk 10 above)
Comment (denis0x0D) [reply lib/Dialect/SPIRV/Serialization/Serializer.cpp]: Thanks, fixed!

File: include/mlir/Dialect/SPIRV/SPIRVTypes.h (hunk 11 above)
Comment (denis0x0D) [reply include/mlir/Dialect/SPIRV/SPIRVTypes.h]: Thanks, updated.

**Comment by denis0x0D** [comment]
@antiagainst thanks a lot for review! I've fixed my incorrect understanding of specifications for ArrayStride, can you please take a look?
Thanks.

**Comment by denis0x0D** [comment]
tests updated.

**Comment by denis0x0D** [comment]
@antiagainst @MaheshRavishankar if no one is interesting in the task https://github.com/tensorflow/mlir/issues/77 , can I take it?
Thanks!

**Comment by antiagainst** [comment]
Sure! Please go ahead!

**Comment by MaheshRavishankar** [comment]
Thanks Denis for taking up the extra tasks. We have been busy with trying to resolve the sticky issue with https://github.com/tensorflow/mlir/issues/68.

**Comment by denis0x0D** [comment]
@antiagainst @MaheshRavishankar thanks, I'm happy to help with it. 
I've seen this issue https://github.com/tensorflow/mlir/issues/68 ,what the solution you've decided to take?

**Comment by MaheshRavishankar** [comment]
@denis0x0D I think we are going to have to change the way module level variables and constants are represented. They will have a symbol name associated with such variables/constants and wont have an SSA value. I am in the process of changing variables to be this way. Will post an update when it goes into github. There are a lot of advantages of not doing implicit captures in functions. But it doesnt match directly with SPIR-V spec, so some things have to change in the current dialect.

**Comment by antiagainst** [comment]
+1 to what @MaheshRavishankar said. We'll update #68 later to capture the offline discussion.

This pull request now has a conflict. Could you rebase against master, @denis0x0D? Thanks!

**Comment by denis0x0D** [comment]
@MaheshRavishankar thanks, sounds interesting!
@antiagainst rebased on master.
Thanks.

**Comment by denis0x0D** [comment]
BTW, it seems like CI build fails, because of hardcoded variable https://github.com/tensorflow/mlir/blob/master/CMakeLists.txt#L10
llvm was moved to C++14 recently and regarding to the https://gcc.gnu.org/projects/cxx-status.html#cxx14 gcc-5 must support c++14 standard.

File: lib/Dialect/SPIRV/SPIRVDialect.cpp [hunk 12]
```diff
@@ -53,6 +53,18 @@ SPIRVDialect::SPIRVDialect(MLIRContext *context)
 // Type Parsing
 //===----------------------------------------------------------------------===//
 
+// Forward declarations.
+template <typename ValTy>
```
Comment (antiagainst) [inline lib/Dialect/SPIRV/SPIRVDialect.cpp]: Could you forward-declare these as `static` to be consistent with later? I'm seeing compiler errors of "static declaration of 'parseAndVerify' follows non-static declaration".

Comment (denis0x0D) [reply lib/Dialect/SPIRV/SPIRVDialect.cpp]: @antiagainst oh, sorry about this, I've updated it, this is indeed a bug regarding to the standard, but I cannot reproduce an error on my local machine with gcc 8, even with reduced test case for template functions.

Comment (antiagainst) [reply lib/Dialect/SPIRV/SPIRVDialect.cpp]: No worries! And thanks for the ping regarding LLVM C++14 update! :)
